Try: Make input styles consistent

jasmussen opened the pull request:
Maybe fixes #4963.

This takes a stab at unifying all basic input fields to inherit and have the same styles. That includes blocks that use input fields, and input fields in the sidebar, for example the alt text on an image block.

Blocks using the PlainText component (which is intentionally unstyled by default) can apply a `regular-text` style that is inherited from upstream. This is now applied to the Shortcode block, which also gets a visual enhancement.

While I believe this PR is a definite enhancement and adds much needed consistency, it should also get many eyes as it unstyles some upstream styles in order to achieve the goal. For example upstream input fields get a glowing blue border, and a drop shadow inside. This unstyles that to achieve consistency in Gutenberg.

As such, before I spend too much time on this, we should decide:

- A) what is the ideal _basic input style_?
- B) once we have that, is it okay to not inherit styles from the rest of WordPress? Should we create an upstream trac ticket to apply Gutenberg input styles to the rest of the admin?

Depending on discussions, I will amend this PR. CC: @paulwilde 

Border styles were also darkened slightly, CC #5482 and @afercia. 

![focusstyles](https://user-images.githubusercontent.com/1204802/37397039-e908466e-277a-11e8-8d90-23ccce58a34f.gif)

Commit:
Try: Make input styles consistent

Maybe fixes #4963.

This takes a stab at unifying all basic input fields to inherit and have the same styles. That includes blocks that use input fields, and input fields in the sidebar, for example the alt text on an image block.

Blocks using the PlainText component (which is intentionally unstyled by default) can apply a `regular-text` style that is inherited from upstream. This is now applied to the Shortcode block, which also gets a visual enhancement.

paulwilde commented:
I'll take a closer look at this tonight most likely.

One thing I've noticed is that you are specifically targeting `input`, surely this would affect checkboxes/radios also?

jasmussen commented:
Thanks. Not urgent.

> One thing I've noticed is that you are specifically targeting input, surely this would affect checkboxes/radios also?

Indeed, I forgot that. This is one of the reasons it's good to test these blanket override styles, and make sure we do it only because it's the right thing to do.

paulwilde commented:
Core specifically targets every input type individually rather than doing a straight `input`. It looks ugly, but it works I guess. One alternative could be to do:

`input:not([type='radio']):not([type='checkbox'])`

https://github.com/WordPress/WordPress/blob/master/wp-admin/css/forms.css#L7

paulwilde commented:
I've tested the branch out and have found a few errors which the highly specific nature of just scoping the global `input` has introduced.

**Range field:**

<img width="261" alt="screen shot 2018-03-17 at 15 18 18" src="https://user-images.githubusercontent.com/704594/37557106-90aea6c8-29f7-11e8-83fc-afe37a461a4a.png">

**Inline link field:**

<img width="323" alt="screen shot 2018-03-17 at 15 34 19" src="https://user-images.githubusercontent.com/704594/37557183-b25c1cb4-29f8-11e8-9dab-b5908e5efa11.png">

**Button link field:**

<img width="310" alt="screen shot 2018-03-17 at 15 34 43" src="https://user-images.githubusercontent.com/704594/37557187-bf36a67a-29f8-11e8-8eb1-4e9e34fa8c3b.png">

**Checkbox fields (These are finally actually, but just for reference on how these look now):**

<img width="275" alt="screen shot 2018-03-17 at 15 41 40" src="https://user-images.githubusercontent.com/704594/37557265-d2d3c3ec-29f9-11e8-9de5-840cdb6765ff.png">


Maybe it's personal preference, or if there's something to consider from a usability/accessibility perspective, however I think I actually preferred the more focal focus style where the input border would turn blue with a bit of box-shadow to make it stand out more. It makes it a lot more obvious where your current input focus is rather than turning the input a slighter darker shade of grey. I kind of like the style which wordpress.com uses. Screenshot for reference:

<img width="269" alt="screen shot 2018-03-17 at 15 30 29" src="https://user-images.githubusercontent.com/704594/37557142-293f0662-29f8-11e8-8d53-de0832735a21.png">

It might be a browser specific thing (I'm using Safari), however I've noticed the line-height of the shortcode field and those in the sidebar seems to differ, and makes the sidebar ones seem quite squashed from the y axis. I'd recommend adding a `line-height: 1.5` or so for the `input, textarea` to make them consistent.

Screenshot of before and after for reference:

<img width="276" alt="screen shot 2018-03-17 at 15 40 49" src="https://user-images.githubusercontent.com/704594/37557245-9a1c038e-29f9-11e8-8f66-fc89b414c945.png">
<img width="275" alt="screen shot 2018-03-17 at 15 40 03" src="https://user-images.githubusercontent.com/704594/37557234-7c2e971a-29f9-11e8-80f1-2e9d534da7bc.png">

jasmussen commented:
Thanks for the feedback. I'll continue to refine this PR (and agree with your notes) 🌟

jasmussen commented:
Okay, pushed some more fixes:

<img width="269" alt="screen shot 2018-03-22 at 11 08 58" src="https://user-images.githubusercontent.com/1204802/37764923-35bdf816-2dc3-11e8-80bd-479cf8878220.png">
<img width="277" alt="screen shot 2018-03-22 at 11 08 54" src="https://user-images.githubusercontent.com/1204802/37764941-3eaf0730-2dc3-11e8-92ed-8351109d6f69.png">
<img width="269" alt="screen shot 2018-03-22 at 11 08 58" src="https://user-images.githubusercontent.com/1204802/37764953-4287ee30-2dc3-11e8-9dd5-93c605f45247.png">

These two are unchanged, though:

<img width="263" alt="screen shot 2018-03-22 at 11 22 21" src="https://user-images.githubusercontent.com/1204802/37764981-555d6846-2dc3-11e8-97f7-2ce6f1cbadf0.png">

<img width="366" alt="screen shot 2018-03-22 at 11 22 17" src="https://user-images.githubusercontent.com/1204802/37764983-57b40622-2dc3-11e8-9f54-999feead58b5.png">

Gutenberg inputs can now opt into the "Gutenberg" input style by applying the class `regular-text`. Excerpt does this, the above two could as well. Long term we'd want to unify the styles between WordPress and Gutenberg, so anything upstream that uses `regular-text` gets the same style. 

But pushing now for discussion and thoughts.

karmatosed commented:
I think this is great and a good step in consistency.

jasmussen commented:
A note to self to maybe also look at improving the color of placeholder text. That is, ensuring that it _looks_ like placeholder text, but still maintaining sufficient contrast. Could mean making non-placeholder text darker.

jasmussen commented:
Squashed and rebased this. Took a look at placeholder text, but turns out we've already styled the placeholder text to be as light a gray as AAA contrast allows ($dark-gray-300 is the lightest we can use).

CC: @ryelle do you know if we can use a lighter-than-AAA for placeholder text? Asking because in an incoming PR, we show _image dimensions_ as placeholder text when unset. For example for a 400x300px image that does _not yet_ have explicit `width` and `height` attributes, we'd show something like (incoming ASCII art):

```
Width: [ 400 ]
Height: [ 300 ]
````

☝️ but because those values are unset, they'd be light gray and `placeholder` values. If we could make those a lighter gray that would make it more clear that they aren't actually set yet. But curious what the contrast requirements are for placeholder text.

paulwilde commented:
Any reason for the class name of `.regular-text`? Something like `.form-control` (Which is what Bootstrap uses) is more obvious its its intentions than `.regular-text` which just sounds like it applies text formatting paragraph.

jasmussen commented:
> Any reason for the class name of .regular-text?

No reason, other than it's the upstream WordPress form field class as far as I can tell.

afercia commented:
As long as all the input fields and textareas have a visible, properly associated `<label>` element OR their accessible name made apparent in some way (e.g. the post title), that's fine from an accessibility perspective.

afercia commented:
@jasmussen on Safari, when I click on the "Document" or "Block" buttons (the "tabs") in the sidebar, i see the buttons text color changes to white while clicking (as far as I can tell it's the Safari default for buttons in the active state). I don't think it's related to this PR though.

paulwilde commented:
> @jasmussen on Safari, when I click on the "Document" or "Block" buttons (the "tabs") in the sidebar, i see the buttons text color changes to white while clicking (as far as I can tell it's the Safari default for buttons in the active state). I don't think it's related to this PR though.

This is unrelated. I've been seeing this since day 1 pretty much. 😄

jasmussen commented:
I've rebased, squashed, polished, fixed tests, and I believe this is ready to go. 

In the last commit, https://github.com/WordPress/gutenberg/pull/5605/commits/ee70ff67369765266ace9d3c7d2df6ea6e259e4a, I also override core styles. We do this because if we don't, we can't fix #4963, because the editor itself relies so on textfields. We could replicate the same styles in the slimmer scope, but as suggested in this thread it might be worth looking at new styles that are more compatible with the general Gutenberg design, and then submit these upstream as well.

Alternately, we can revert that last commit.

Screenshots:

![screen shot 2018-04-11 at 15 36 48](https://user-images.githubusercontent.com/1204802/38620759-a8fab2c4-3d9f-11e8-9169-e42c01980dab.png)

![screen shot 2018-04-11 at 15 36 59](https://user-images.githubusercontent.com/1204802/38620763-aa83c3d8-3d9f-11e8-82e1-f420581d58c1.png)

![screen shot 2018-04-11 at 15 37 08](https://user-images.githubusercontent.com/1204802/38620765-abef6bfa-3d9f-11e8-9057-3847a8260c8c.png)

![screen shot 2018-04-11 at 15 43 06](https://user-images.githubusercontent.com/1204802/38620767-ad48f5a2-3d9f-11e8-86b1-e29ca50eb0ad.png)

![screen shot 2018-04-11 at 15 43 31](https://user-images.githubusercontent.com/1204802/38620772-af00ba4c-3d9f-11e8-8b65-d770c57e0992.png)

@paulwilde another look? Thank you!

paulwilde commented:
@jasmussen Just tried it out, looks good.

If I had any nitpicks, it would be to change the focus outline to $blue-medium-600 (#00a0d2) as the grey against the light blue looks a little dull to me. However I don't mind too much either way. 

Additionally I think the left/right padding of 10px is a little too deep, I think reducing it down 1 or 2 pixels looks better. It also then keeps it consistent with `.components-text-control__input` which has them set as 8px. Also core doesn't have deep padding either (3px 5px vs 6px 10px).

Before:
<img width="304" alt="screen shot 2018-04-12 at 15 16 40" src="https://user-images.githubusercontent.com/704594/38683272-f929d750-3e64-11e8-8bb5-68acab47d789.png">

After:
<img width="296" alt="screen shot 2018-04-12 at 15 21 39" src="https://user-images.githubusercontent.com/704594/38683382-391925fa-3e65-11e8-87d9-ea6e7160c99d.png">

Commit:
Adjust focus color, and padding.

paulwilde commented:
Sorry, by outline I mean the 1px border around the actual input when its in focus. If you compare the hex between my 2 images you should notice what I mean.

Commit:
Address feedback.

jasmussen commented:
Awesome, looks good:

<img width="283" alt="screen shot 2018-04-13 at 11 47 37" src="https://user-images.githubusercontent.com/1204802/38728564-838f66d4-3f10-11e8-9419-56e459ca9cdd.png">

Thank you for the reviews, merging when the checks pass!

jasmussen commented:
🎉
